{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## MyVariant.info and MyGene.info Use Case"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The following R script demonstrates the utility of MyVariant.info and MyGene.info R clients to annotate variants and prioritize candidate genes in patients with rare Mendelian diseases. This specific study uses data obtained from the database of phenotype and genotype (dbGaP) study...FASTQ files generated by Ng et al for the Miller syndrome study. FASTQ files were processed according to the Broad Institute’s best practices. Individual samples were aligned to the hg19 reference genome using BWA-MEM 0.7.10. Variants were called using GATK 3.3-0 HaplotypeCaller and quality scores were recalibrated using GATK VariantRecalibrator."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "The following objects are masked from ‘package:parallel’:\n",
      "\n",
      "    clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,\n",
      "    clusterExport, clusterMap, parApply, parCapply, parLapply,\n",
      "    parLapplyLB, parRapply, parSapply, parSapplyLB\n",
      "\n",
      "The following object is masked from ‘package:stats’:\n",
      "\n",
      "    xtabs\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, append, as.data.frame, as.vector, cbind, colnames,\n",
      "    do.call, duplicated, eval, evalq, Filter, Find, get, intersect,\n",
      "    is.unsorted, lapply, Map, mapply, match, mget, order, paste, pmax,\n",
      "    pmax.int, pmin, pmin.int, Position, rank, rbind, Reduce, rep.int,\n",
      "    rownames, sapply, setdiff, sort, table, tapply, union, unique,\n",
      "    unlist, unsplit\n",
      "\n",
      "Creating a generic function for ‘nchar’ from package ‘base’ in package ‘S4Vectors’\n",
      "\n",
      "Attaching package: ‘VariantAnnotation’\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    tabulate\n",
      "\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "Loading required package: DBI\n",
      "\n",
      "\n",
      "Attaching package: ‘plyr’\n",
      "\n",
      "The following object is masked from ‘package:XVector’:\n",
      "\n",
      "    compact\n",
      "\n",
      "The following object is masked from ‘package:IRanges’:\n",
      "\n",
      "    desc\n",
      "\n",
      "The following object is masked from ‘package:S4Vectors’:\n",
      "\n",
      "    rename\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(myvariant, quietly=TRUE)\n",
    "library(mygene, quietly=TRUE)\n",
    "library(VariantAnnotation, quietly=TRUE)\n",
    "library(GO.db, quietly=TRUE)\n",
    "source(\"https://raw.githubusercontent.com/Network-of-BioThings/myvariant.info/master/docs/ipynb/mendelian.R\")\n",
    "setwd(\"~/sulab/myvariant/vcf/recal\")\n",
    "vcf.files <- paste(getwd(), list.files(getwd()), sep=\"/\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "mendelian.R defines some helper functions that are used in the analysis downstream of retrieving annotations:\n",
    "\n",
    "`replaceWith0` - replaces all NAs in a data.frame with 0.\n",
    "\n",
    "`rankByCaddScore` - for prioritizing genes by deleteriousness (scaled CADD score)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "## Annotating variants with MyVariant.info\n",
    "The following function reads in each output VCF file using the VariantAnnotation package available from Bioconductor. Install with `biocLite(\"VariantAnnotation\")`. `formatHgvs` (from the ***myvariant*** Bioconductor package) is a function that reads the genomic location and variant information from the VCF to create HGVS IDs which serve as a primary key for each variant. The function `getVariants` makes the queries to MyVariant.info to retrieve annotations."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing  /Users/cyrusafrasiabi/recal/subject01_recalibrate_SNP_vqsr.vcf ...\n",
      "found header lines for 3 ‘fixed’ fields: ALT, QUAL, FILTER \n",
      "found header lines for 24 ‘info’ fields: AC, AF, ..., VQSLOD, culprit \n",
      "found header lines for 5 ‘geno’ fields: GT, AD, DP, GQ, PL \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Concatenating data, please be patient.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing  /Users/cyrusafrasiabi/recal/subject02_recalibrate_SNP_vqsr.vcf ...\n",
      "found header lines for 3 ‘fixed’ fields: ALT, QUAL, FILTER \n",
      "found header lines for 24 ‘info’ fields: AC, AF, ..., VQSLOD, culprit \n",
      "found header lines for 5 ‘geno’ fields: GT, AD, DP, GQ, PL \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Concatenating data, please be patient.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing  /Users/cyrusafrasiabi/recal/subject03_recalibrate_SNP_vqsr.vcf ...\n",
      "found header lines for 3 ‘fixed’ fields: ALT, QUAL, FILTER \n",
      "found header lines for 24 ‘info’ fields: AC, AF, ..., VQSLOD, culprit \n",
      "found header lines for 5 ‘geno’ fields: GT, AD, DP, GQ, PL \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Concatenating data, please be patient.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing  /Users/cyrusafrasiabi/recal/subject04_recalibrate_SNP_vqsr.vcf ...\n",
      "found header lines for 3 ‘fixed’ fields: ALT, QUAL, FILTER \n",
      "found header lines for 24 ‘info’ fields: AC, AF, ..., VQSLOD, culprit \n",
      "found header lines for 5 ‘geno’ fields: GT, AD, DP, GQ, PL \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Concatenating data, please be patient.\n"
     ]
    }
   ],
   "source": [
    "getVars <- function(vcf.file){\n",
    "  cat(paste(\"Processing \", vcf.file, \"...\\n\", sep=\" \"))\n",
    "  vcf <- readVcf(vcf.file, genome=\"hg19\")\n",
    "  vcf <- vcf[isSNV(vcf)]\n",
    "  vars <- rowRanges(vcf)\n",
    "  vars <- as(vars, \"DataFrame\")\n",
    "  vars$query <- formatHgvs(vcf, \"snp\")\n",
    "  annotations <- getVariants(vars$query, fields=c(\"dbnsfp.genename\", \"dbnsfp.1000gp1.af\", \n",
    "                                                  \"exac.af\", \"cadd.consequence\", \"cadd.phred\"), verbose=FALSE)\n",
    "  annotations[c('DP', 'FS', 'QD')] <- info(vcf)[c('DP', 'FS', 'QD')]\n",
    "  annotations <- replaceWith0(annotations)\n",
    "  annotations\n",
    "}\n",
    "\n",
    "vars <- lapply(vcf.files, getVars)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We then filter each of the VCF files based on annotations."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2442 common genes are mutated amongst patients"
     ]
    }
   ],
   "source": [
    "countNumGenesMutatedInAllPatients <- function(inp, text) {\n",
    "    cat(paste(nrow(subset(data.frame(table(unlist(lapply(inp, function(i) unique(i$dbnsfp.genename))))), \n",
    "    Freq == 4)), text))\n",
    "}\n",
    "                                                         \n",
    "countNumGenesMutatedInAllPatients(vars, \"common genes are mutated amongst patients\")                                                         "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "## Filtering Steps\n",
    "The next steps apply annotation based filters by subsetting each data.frame by those variants that meet each criteria.\n",
    "\n",
    "1) The first filter utilizes annotations output by HaplotypeCaller from GATK. Here we keep variants with at least 8 reads, Fisher Strand bias less than 30, and quality over depth greater than 2. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2309 genes remain after filtering for coverage and strand bias"
     ]
    }
   ],
   "source": [
    "filter1 <- lapply(vars, function(i) subset(i, DP > 8 & FS < 30 & QD > 2))\n",
    "\n",
    "countNumGenesMutatedInAllPatients(filter1, \"genes remain after filtering for coverage and strand bias\") "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "2) Mendelian diseases are most likely to be caused by nonsynonymous mutations. We can take advantage of the CADD database that denotes the mutation type in the field \"cadd.consequence\"."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1918 genes remain after filtering for nonsynonymous and splice site variants"
     ]
    }
   ],
   "source": [
    "filter2 <- lapply(filter1, function(i) subset(i, cadd.consequence %in% c(\"NON_SYNONYMOUS\", \"STOP_GAINED\", \"STOP_LOST\", \n",
    "                                           \"CANONICAL_SPLICE\", \"SPLICE_SITE\")))\n",
    "\n",
    "countNumGenesMutatedInAllPatients(filter2, \"genes remain after filtering for nonsynonymous and splice site variants\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "3) The third filter keeps rare variants according to the ExAC data set with allele frequency < 0.01. Rare diseases are likely caused by mutations that have not been documented yet."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "19 genes remain after filtering for allele frequency < 0.01 in the ExAC dataset"
     ]
    }
   ],
   "source": [
    "filter3 <- lapply(filter2, function(i) subset(i, exac.af < 0.01))\n",
    "\n",
    "countNumGenesMutatedInAllPatients(filter3, \"genes remain after filtering for allele frequency < 0.01 in the ExAC dataset\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "4) The last filter keeps rare variants according to the 1000 Genomes Project with allele frequency < 0.01."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "9 genes remain after filtering for allele frequency < 0.01 in 1000 Genomes Project"
     ]
    }
   ],
   "source": [
    "filter4 <- lapply(filter3, function(i) subset(i, sapply(dbnsfp.1000gp1.af, function(j) j < 0.01 )))\n",
    "    \n",
    "top.genes <- subset(data.frame(table(unlist(lapply(filter4, function(i) unique(i$dbnsfp.genename))))), Freq == 4)\n",
    "top.genes <- subset(top.genes, !(Var1 %in% c(\"NULL\", 0)))\n",
    "cat(paste(nrow(top.genes), \"genes remain after filtering for allele frequency < 0.01 in 1000 Genomes Project\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<ol class=list-inline>\n",
       "\t<li>CDC27</li>\n",
       "\t<li>CDON</li>\n",
       "\t<li>CTBP2</li>\n",
       "\t<li>DHODH</li>\n",
       "\t<li>FAM104B</li>\n",
       "\t<li>HYDIN</li>\n",
       "\t<li>KIR3DL3</li>\n",
       "\t<li>PIK3R3</li>\n",
       "\t<li>XKR3</li>\n",
       "</ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item CDC27\n",
       "\\item CDON\n",
       "\\item CTBP2\n",
       "\\item DHODH\n",
       "\\item FAM104B\n",
       "\\item HYDIN\n",
       "\\item KIR3DL3\n",
       "\\item PIK3R3\n",
       "\\item XKR3\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. CDC27\n",
       "2. CDON\n",
       "3. CTBP2\n",
       "4. DHODH\n",
       "5. FAM104B\n",
       "6. HYDIN\n",
       "7. KIR3DL3\n",
       "8. PIK3R3\n",
       "9. XKR3\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] CDC27   CDON    CTBP2   DHODH   FAM104B HYDIN   KIR3DL3 PIK3R3  XKR3   \n",
       "968 Levels: 0 ABCA2 ABCA6 ABCB6 ABCC10 ABCC11 ABCC5 ABHD5 ABLIM3 ABTB1 ... ZWINT"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "top.genes$Var1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "## Prioritizing genes\n",
    "We can prioritize the genes that contain variants in each of the patients according to CADD (deleteriousness) score. `rankByCaddScore` extracts the average CADD scores of the variants in each gene per patient and ranks in descending order."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<thead><tr><th></th><th scope=col>gene</th><th scope=col>cadd.phred</th></tr></thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>DHODH</td><td>26.81</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>CTBP2</td><td>21.385</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>PIK3R3</td><td>20.7</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>HYDIN</td><td>18.91</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>CDC27</td><td>18.545</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>XKR3</td><td>15.72</td></tr>\n",
       "\t<tr><th scope=row>7</th><td>CDON</td><td>10.02</td></tr>\n",
       "\t<tr><th scope=row>8</th><td>FAM104B</td><td>7.2055</td></tr>\n",
       "\t<tr><th scope=row>9</th><td>KIR3DL3</td><td>1.012</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "\\begin{tabular}{r|ll}\n",
       "  & gene & cadd.phred\\\\\n",
       "\\hline\n",
       "\t1 & DHODH & 26.81\\\\\n",
       "\t2 & CTBP2 & 21.385\\\\\n",
       "\t3 & PIK3R3 & 20.7\\\\\n",
       "\t4 & HYDIN & 18.91\\\\\n",
       "\t5 & CDC27 & 18.545\\\\\n",
       "\t6 & XKR3 & 15.72\\\\\n",
       "\t7 & CDON & 10.02\\\\\n",
       "\t8 & FAM104B & 7.2055\\\\\n",
       "\t9 & KIR3DL3 & 1.012\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/plain": [
       "     gene cadd.phred\n",
       "1   DHODH    26.8100\n",
       "2   CTBP2    21.3850\n",
       "3  PIK3R3    20.7000\n",
       "4   HYDIN    18.9100\n",
       "5   CDC27    18.5450\n",
       "6    XKR3    15.7200\n",
       "7    CDON    10.0200\n",
       "8 FAM104B     7.2055\n",
       "9 KIR3DL3     1.0120"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ranked <- rankByCaddScore(top.genes$Var1, filter4)\n",
    "ranked"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "##Annotating candidate genes with MyGene.info\n",
    "Now we can annotate the genes using the `queryMany` function from ***mygene*** which can be install from Bioconductor (`biocLite(\"mygene\")`). We specify `scopes=symbol` and request annotations by the `field` parameter. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Finished\n"
     ]
    }
   ],
   "source": [
    "res <- data.frame(queryMany(ranked$gene, scopes=\"symbol\", species=\"human\", fields=c(\"go.BP\", \"name\", \"MIM\", \"uniprot\")))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[[1]]\n",
       "   evidence                                                   term         id\n",
       "1       TAS                pyrimidine nucleobase metabolic process GO:0006206\n",
       "2       IBA   'de novo' pyrimidine nucleobase biosynthetic process GO:0006207\n",
       "3       IEA                                       female pregnancy GO:0007565\n",
       "4       IEA                                              lactation GO:0007595\n",
       "5       IEA                                   response to caffeine GO:0031000\n",
       "6       IEA                                       response to drug GO:0042493\n",
       "7       IEA                                 response to starvation GO:0042594\n",
       "8       IEA               positive regulation of apoptotic process GO:0043065\n",
       "9       IEA                     'de novo' UMP biosynthetic process GO:0044205\n",
       "10      TAS                       small molecule metabolic process GO:0044281\n",
       "11      TAS             pyrimidine nucleoside biosynthetic process GO:0046134\n",
       "12      TAS nucleobase-containing small molecule metabolic process GO:0055086\n",
       "13      IEA                            oxidation-reduction process GO:0055114\n",
       "14      IEA                    regulation of mitochondrial fission GO:0090140\n",
       "\n",
       "[[2]]\n",
       "   evidence\n",
       "1       IBA\n",
       "2       IEA\n",
       "3       TAS\n",
       "4       TAS\n",
       "5       IEA\n",
       "6       ISS\n",
       "7       IMP\n",
       "8       IMP\n",
       "9       ISS\n",
       "10      IEA\n",
       "                                                                   term\n",
       "1  negative regulation of transcription from RNA polymerase II promoter\n",
       "2                                          transcription, DNA-templated\n",
       "3                             negative regulation of cell proliferation\n",
       "4                                              viral genome replication\n",
       "5                              positive regulation of chromatin binding\n",
       "6                   negative regulation of transcription, DNA-templated\n",
       "7  positive regulation of transcription from RNA polymerase II promoter\n",
       "8       positive regulation of retinoic acid receptor signaling pathway\n",
       "9                                        white fat cell differentiation\n",
       "10                                          oxidation-reduction process\n",
       "           id   pubmed\n",
       "1  GO:0000122       NA\n",
       "2  GO:0006351       NA\n",
       "3  GO:0008285  9479502\n",
       "4  GO:0019079  9479502\n",
       "5  GO:0035563       NA\n",
       "6  GO:0045892       NA\n",
       "7  GO:0045944 23775127\n",
       "8  GO:0048386 23775127\n",
       "9  GO:0050872       NA\n",
       "10 GO:0055114       NA\n",
       "\n",
       "[[3]]\n",
       "  evidence\n",
       "1      TAS\n",
       "2      TAS\n",
       "3      TAS\n",
       "4      TAS\n",
       "5      IBA\n",
       "6      TAS\n",
       "7      IBA\n",
       "8      ISS\n",
       "                                                                   term\n",
       "1                                        phospholipid metabolic process\n",
       "2                             phosphatidylinositol biosynthetic process\n",
       "3                                    insulin receptor signaling pathway\n",
       "4                 phosphatidylinositol-3-phosphate biosynthetic process\n",
       "5                  regulation of phosphatidylinositol 3-kinase activity\n",
       "6                                      small molecule metabolic process\n",
       "7                                  phosphatidylinositol phosphorylation\n",
       "8 positive regulation of glucose import in response to insulin stimulus\n",
       "          id  pubmed\n",
       "1 GO:0006644      NA\n",
       "2 GO:0006661      NA\n",
       "3 GO:0008286 9524259\n",
       "4 GO:0036092 9524259\n",
       "5 GO:0043551      NA\n",
       "6 GO:0044281      NA\n",
       "7 GO:0046854      NA\n",
       "8 GO:2001275      NA\n",
       "\n",
       "[[4]]\n",
       "  evidence                           term         id\n",
       "1      IEA    epithelial cell development GO:0002064\n",
       "2      IEA                cilium movement GO:0003341\n",
       "3      IEA ventricular system development GO:0021591\n",
       "4      IEA               axoneme assembly GO:0035082\n",
       "5      IEA            trachea development GO:0060438\n",
       "\n",
       "[[5]]\n",
       "  evidence\n",
       "1      TAS\n",
       "2      IMP\n",
       "3      TAS\n",
       "4      TAS\n",
       "5      TAS\n",
       "6      TAS\n",
       "7      TAS\n",
       "8      TAS\n",
       "9      IDA\n",
       "                                                                                                              term\n",
       "1                                                                                               mitotic cell cycle\n",
       "2                                                              metaphase/anaphase transition of mitotic cell cycle\n",
       "3                                                                              mitotic spindle assembly checkpoint\n",
       "4                                                                                               cell proliferation\n",
       "5                   anaphase-promoting complex-dependent proteasomal ubiquitin-dependent protein catabolic process\n",
       "6                          negative regulation of ubiquitin-protein ligase activity involved in mitotic cell cycle\n",
       "7 positive regulation of ubiquitin-protein ligase activity involved in regulation of mitotic cell cycle transition\n",
       "8                                   regulation of ubiquitin-protein ligase activity involved in mitotic cell cycle\n",
       "9                                                                                protein K11-linked ubiquitination\n",
       "          id   pubmed\n",
       "1 GO:0000278       NA\n",
       "2 GO:0007091  7736578\n",
       "3 GO:0007094       NA\n",
       "4 GO:0008283  7736578\n",
       "5 GO:0031145       NA\n",
       "6 GO:0051436       NA\n",
       "7 GO:0051437       NA\n",
       "8 GO:0051439       NA\n",
       "9 GO:0070979 18485873\n",
       "\n",
       "[[6]]\n",
       "NULL\n",
       "\n",
       "[[7]]\n",
       "   evidence\n",
       "1       IEA\n",
       "2       IEA\n",
       "3       TAS\n",
       "4       IEA\n",
       "5       IEA\n",
       "6       IEA\n",
       "7       IEA\n",
       "8       IEA\n",
       "9       TAS\n",
       "10      IEA\n",
       "11      IEA\n",
       "12      IEA\n",
       "13      IEA\n",
       "14      IEA\n",
       "15      IEA\n",
       "16      TAS\n",
       "17      IEA\n",
       "18      IEA\n",
       "                                                                   term\n",
       "1                                               cell fate specification\n",
       "2                                   lens development in camera-type eye\n",
       "3                                                         cell adhesion\n",
       "4                                                       myoblast fusion\n",
       "5                              anterior/posterior pattern specification\n",
       "6                                          embryonic body morphogenesis\n",
       "7                        skeletal muscle satellite cell differentiation\n",
       "8                                           cerebral cortex development\n",
       "9                                           muscle cell differentiation\n",
       "10                                  positive regulation of MAPK cascade\n",
       "11                    regulation of protein heterodimerization activity\n",
       "12                        positive regulation of neuron differentiation\n",
       "13 positive regulation of transcription from RNA polymerase II promoter\n",
       "14            positive regulation of skeletal muscle tissue development\n",
       "15     positive regulation of small GTPase mediated signal transduction\n",
       "16                   positive regulation of muscle cell differentiation\n",
       "17                    embryonic retina morphogenesis in camera-type eye\n",
       "18           positive regulation of neural precursor cell proliferation\n",
       "           id  pubmed\n",
       "1  GO:0001708      NA\n",
       "2  GO:0002088      NA\n",
       "3  GO:0007155 9214393\n",
       "4  GO:0007520      NA\n",
       "5  GO:0009952      NA\n",
       "6  GO:0010172      NA\n",
       "7  GO:0014816      NA\n",
       "8  GO:0021987      NA\n",
       "9  GO:0042692      NA\n",
       "10 GO:0043410      NA\n",
       "11 GO:0043497      NA\n",
       "12 GO:0045666      NA\n",
       "13 GO:0045944      NA\n",
       "14 GO:0048643      NA\n",
       "15 GO:0051057      NA\n",
       "16 GO:0051149      NA\n",
       "17 GO:0060059      NA\n",
       "18 GO:2000179      NA\n",
       "\n",
       "[[8]]\n",
       "NULL\n",
       "\n",
       "[[9]]\n",
       "NULL\n"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res$go.BP"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "# GO annotation\n",
    "We can then use the Bioconductor package ***GO.db*** to map GO IDs to GO:0008152, the biological process id for metabolic process."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<ol class=list-inline>\n",
       "\t<li>DHODH</li>\n",
       "\t<li>CTBP2</li>\n",
       "\t<li>PIK3R3</li>\n",
       "\t<li>CDC27</li>\n",
       "\t<li>CDON</li>\n",
       "</ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item DHODH\n",
       "\\item CTBP2\n",
       "\\item PIK3R3\n",
       "\\item CDC27\n",
       "\\item CDON\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. DHODH\n",
       "2. CTBP2\n",
       "3. PIK3R3\n",
       "4. CDC27\n",
       "5. CDON\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] DHODH  CTBP2  PIK3R3 CDC27  CDON  \n",
       "Levels: CDC27 CDON CTBP2 DHODH FAM104B HYDIN KIR3DL3 PIK3R3 XKR3"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "miller.bp <- lapply(res$go.BP, function(i) unlist(i$id))\n",
    "bp.ancestor <- lapply(miller.bp, function(i) sapply(i, function(j) \"GO:0008152\" %in% unlist(GOBPANCESTOR[[j]])))\n",
    "candidate.genes <- ranked$gene[sapply(bp.ancestor, function(i) TRUE %in% i)]\n",
    "candidate.genes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "The final candidate gene list contains only genes involved in metabolic processes where perterbation is more likely to result in mendelian disease. Through the use of MyVariant.info and MyGene.info annotation services, we have narrowed the candidate gene list from 2589 genes to seven. While we cannot fully remove the process of biological interpretation from an analysis, we have greatly reduced the burden."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.2.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
